### PR TITLE
feat: add token-budget based file reading with intelligent preview

### PR DESCRIPTION
**Pull request by daniel-lxs**
## Problem

Files can fill the entire context window, and tiktoken crashes with `RuntimeError: unreachable` on files >5MB.
## Solution

Token-budget based file reading with multi-layer protection against context overflow and tokenizer crashes.

## How It Works

```
File Read Request
       ↓
Check file size
       ↓
   < 100KB? → Read normally (fast path)
       ↓
   > 5MB? → Preview first 100KB (crash prevention)
       ↓
Count tokens → Within budget? → Full content
                     ↓
                Exceeds budget? → Truncate to fit
```

## Implementation

**Layer 1: Fast Path (< 100KB)**
- Skip validation for small files - zero overhead ⚡

**Layer 2: Token Validation (100KB - 5MB)**
- Dynamic budget: `(contextWindow - currentTokens) * 0.6`
- Real token counting, smart truncation if needed 🎯

**Layer 3: Preview Mode (> 5MB)**
- Return 100KB preview to prevent crashes
- Suggests using `line_range` for targeted reading 👁️

**Layer 4: Error Recovery**
- Catch tokenizer `unreachable` errors gracefully
- Fallback to 100KB preview instead of crashing 🛡️

## Benefits

✅ Dynamic budget based on actual context (no magic numbers)  
✅ Real token counting using existing tokenizer  
✅ 100KB previews for large files  
✅ Graceful error handling prevents conversation crashes  
✅ Simple (~160 lines) vs complex heuristics  
✅ 17 comprehensive tests covering all scenarios

## Testing

All 17 tests passing: fast path, budget validation, preview mode, error recovery, edge cases, unicode support.

## Related

Closes #6667

----

> [!IMPORTANT]
> Introduces token-budget based file reading in `readFileTool.ts` to handle large files efficiently, with new functions for budget validation and content truncation, and comprehensive tests.
> 
>   - **Behavior**:
>     - Implements token-budget based file reading in `readFileTool.ts` to prevent context overflow and tokenizer crashes.
>     - Introduces `validateFileTokenBudget` and `truncateFileContent` in `fileTokenBudget.ts` for handling large files.
>     - Handles files <100KB normally, previews first 100KB for files >5MB, and truncates content if it exceeds token budget.
>   - **Testing**:
>     - Adds `fileTokenBudget.spec.ts` with 17 tests covering scenarios like small files, large files, budget validation, and error handling.
>     - Updates `line-counter.spec.ts` to test line and token counting with error handling.
>   - **Misc**:
>     - Updates `line-counter.ts` to include token estimation alongside line counting.
>     - Modifies `readFileTool.spec.ts` to test new file reading behavior.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 169fb35786a49b10db5ad86705e585c604bfc27b. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by ghost** [comment]
Starting my review of this PR—comments incoming soon! 🚀

[Follow Along on Roo Code Cloud](https://app.roocode.com/cloud-jobs/UlQrYszT?utm_source=cloud-job-url&utm_medium=link&utm_campaign=github.pr.review)

---

## Issues to Address

All previously flagged issues have been resolved:

- [x] The `lines` attribute shows `totalLines` from the original file even after content is truncated (see inline comment on readFileTool.ts:619)
- [x] Empty lines are being filtered out when counting displayed lines, causing incorrect line count in the `lines` attribute (see inline comment on readFileTool.ts:620-621)
- [x] Line counting logic has off-by-one error for files ending with newline characters (see inline comment on readFileTool.ts:620)

**Latest Review:** No new issues found. All previous concerns have been properly addressed.

File: src/core/tools/readFileTool.ts [hunk 1]
```diff
@@ -594,13 +595,40 @@ export async function readFileTool(
...
+						content,
+						budgetResult.maxChars,
+						content.length,
+						budgetResult.isPreview,
+					)
+					content = truncateResult.content
+
+					// Reflect actual displayed line count after truncation
+					const displayedLines =
+						content.length > 0 ? content.split(/\r?\n/).filter((l) => l !== "").length || 1 : 0
```
Comment (ghost) [inline src/core/tools/readFileTool.ts]: Empty lines are being filtered out when counting displayed lines, causing the line count to be incorrect. If truncated content contains empty lines, they won't be counted, making the `lines` attribute claim fewer lines than are actually displayed. For example, if the truncated content has 100 lines including 20 empty lines, this will report `lines="1-80"` instead of `lines="1-100"`. Empty lines are valid lines in source code and should be counted in the range.

```suggestion
const displayedLines =
	content.length > 0 ? content.split(/\r?\n/).length : 0
```

File: src/core/tools/helpers/fileTokenBudget.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+			// Count tokens with error handling
+			try {
+				const contentBlocks: Anthropic.Messages.ContentBlockParam[] = [{ type: "text", text: content }]
+				tokenCount = await countTokens(contentBlocks)
+			} catch (error) {
+				// Catch tokenizer "unreachable" errors
+				const errorMessage = error instanceof Error ? error.message : String(error)
+				if (errorMessage.includes("unreachable")) {
+					// Use conservative estimation: 2 chars = 1 token
+					const estimatedTokens = Math.ceil(content.length / 2)
```
[the remark this answers is not in the document]
Comment (daniel-lxs) [reply src/core/tools/helpers/fileTokenBudget.ts:149]: good catch!

File: src/core/tools/readFileTool.ts [hunk 3]
```diff
@@ -594,13 +595,39 @@ export async function readFileTool(
...
+					const truncateResult = truncateFileContent(
+						content,
+						budgetResult.maxChars,
+						content.length,
+						budgetResult.isPreview,
+					)
+					content = truncateResult.content
+
+					// Reflect actual displayed line count after truncation (count ALL lines, including empty)
+					const displayedLines = content.length === 0 ? 0 : content.split(/\r?\n/).length
```
[the remark this answers is not in the document]
Comment (daniel-lxs) [reply src/core/tools/readFileTool.ts]: Done

**Review by mrubens — APPROVED** [review]
